### PR TITLE
Supporting negative axes for all existing onnx ops

### PR DESCRIPTION
**Pull request by neginraoof**
I modified the ops documentations as well as tests, to add support for negative axes

### CODE REVIEW

**Comment by wschin** [comment]
What's the relation between https://github.com/onnx/onnx/pull/2207 and this PR?

**Comment by neginraoof** [comment]
@wschin I just saw this PR. Looks like it does not increase the version number, and it' missing tests for negative axes. This PR also includes most recent ops. I can merge that PR and make sure those fixes are included here.

**Comment by skottmckay** [comment]
> @wschin I just saw this PR. Looks like it does not increase the version number, and it' missing tests for negative axes. This PR also includes most recent ops. I can merge that PR and make sure those fixes are included here.

Please do. Can you also rev the Split opset for consistency? Negative axis support was added there recently without that happening (https://github.com/onnx/onnx/pull/2177).

**Comment by neginraoof** [comment]
cc @BowenBao @spandantiwari to review

File: onnx/defs/operator_sets.h [hunk 1]
```diff
@@ -620,6 +644,30 @@ class OpSet_Onnx_ver11 {
     fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, Det)>());
     fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, ScatterND)>());
     fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, GatherND)>());
+    fn(GetOpSchema<ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 11, Slice)>());
```
Comment (spandantiwari) [inline onnx/defs/operator_sets.h]: An FYI comment for all - It was discussed during the SIG/Operators WG in ONNX workshop that it is safer option to bump up the opset version of these ops now that the spec is clear that negative axes are allowed.

File: docs/Changelog.md [hunk 2]
```diff
@@ -1204,7 +1204,7 @@ This version of the operator has been available since version 1 of the default O
 
 <dl>
 <dt><tt>axis</tt> : int (default is 0)</dt>
-<dd>Which axis to gather on. Negative value means counting dimensions from the back. Accepted range in [-r, r-1]</dd>
+<dd>Which axis to gather on. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: > r = rank(input). [](start = 118, length = 17)

Very nice comment.

File: docs/Changelog.md [hunk 3]
```diff
@@ -4739,7 +4739,7 @@ This version of the operator has been available since version 2 of the default O
 
 <dl>
 <dt><tt>axis</tt> : int (default is 0)</dt>
-<dd>Which axis to split on. A negative value means counting dimensions from the back. Accepted range is [-rank, rank-1].</dd>
```
Comment (wschin) [inline docs/Changelog.md]: > A negative value means counting dimensions from the back. Accepted range is [-rank, rank-1]. [](start = 27, length = 93)

If the old description allows negative axis, the old signature should still allow it.

File: docs/Changelog.md [hunk 4]
```diff
@@ -10286,6 +10286,88 @@ This version of the operator has been available since version 10 of the default
...
+
+#### Version
+
+This version of the operator has been available since version 11 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>axis</tt> : int (default is 0)</dt>
+<dd>The axis in which to compute the arg indices. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 93, length = 5)

input-->data
Please use the tensor names below.

File: docs/Changelog.md [hunk 5]
```diff
@@ -10286,6 +10286,88 @@ This version of the operator has been available since version 10 of the default
...
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 93, length = 5)

input-->data

File: docs/Changelog.md [hunk 6]
```diff
@@ -10370,6 +10452,85 @@ This version of the operator has been available since version 11 of the default
...
+
+#### Version
+
+This version of the operator has been available since version 11 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>axis</tt> : int (required)</dt>
+<dd>Which axis to concat on. A negative value means counting dimensions from the back. Accepted range is [-rank, rank-1].</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >[-rank, rank-1] [](start = 105, length = 15)

where rank is the rank of inputs.

File: docs/Changelog.md [hunk 7]
```diff
@@ -10725,7 +10924,7 @@ This version of the operator has been available since version 11 of the default
 
 <dl>
 <dt><tt>axis</tt> : int (default is 0)</dt>
-<dd>Which axis to gather on. Negative value means counting dimensions from the back. Accepted range in [-r, r-1]</dd>
+<dd>Which axis to gather on. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 128, length = 5)

data-->input.
Please also double-check other places. Thank you.

File: docs/Changelog.md [hunk 8]
```diff
@@ -11004,6 +11303,61 @@ This version of the operator has been available since version 11 of the default
 <dd>tensor of bool, which should be a scalar.</dd>
 </dl>
 
+### <a name="OneHot-11"></a>**OneHot-11**</a>
```
Comment (wschin) [inline docs/Changelog.md]: >OneHot-11"></a>**OneHot-11**</a> [](start = 13, length = 32)

We don't need to bump the version of OneHot because the default "axis" value in OneHot-9 is already "-1". It means negative axis is explicitly allowed.

File: docs/Changelog.md [hunk 9]
```diff
@@ -11078,110 +11435,705 @@ This version of the operator has been available since version 11 of the default
 #### Attributes
 
 <dl>
-<dt><tt>coordinate_transformation_mode</tt> : string (default is half_pixel)</dt>
-<dd>
-This attribute describes how to transform the coordinate in the resized tensor to the coordinate in the original tensor. <br/>
+<dt><tt>axes</tt> : list of ints</dt>
+<dd>A list of integers, along which to reduce. The default is to reduce over all the dimensions of the input tensor. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 160, length = 5)

input->data

File: docs/Changelog.md [hunk 10]
```diff
@@ -11078,110 +11435,705 @@ This version of the operator has been available since version 11 of the default
...
+
+#### Version
+
+This version of the operator has been available since version 11 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>axes</tt> : list of ints</dt>
+<dd>A list of integers, along which to reduce. The default is to reduce over all the dimensions of the input tensor. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 160, length = 5)

data

File: docs/Changelog.md [hunk 11]
```diff
@@ -11078,110 +11435,705 @@ This version of the operator has been available since version 11 of the default
...
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 160, length = 5)

input->data.
Please double-check all reduction's signatures. Thanks.

File: docs/Changelog.md [hunk 12]
```diff
@@ -11306,7 +12258,7 @@ This version of the operator has been available since version 11 of the default
 
 <dl>
 <dt><tt>axis</tt> : int (default is 0)</dt>
-<dd>Which axis to scatter on. Negative value means counting dimensions from the back. Accepted range in [-r, r-1]</dd>
+<dd>Which axis to scatter on. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 129, length = 5)

data

File: docs/Changelog.md [hunk 13]
```diff
@@ -11424,6 +12376,207 @@ This version of the operator has been available since version 11 of the default
...
+
+#### Type Constraints
+
+<dl>
+<dt><tt>T</tt> : tensor(float16), tensor(float), tensor(double)</dt>
+<dd>Constrain input and output types to float tensors.</dd>
+</dl>
+
+### <a name="Split-11"></a>**Split-11**</a>
+
```
Comment (wschin) [inline docs/Changelog.md]: It's really nice that you discarded the change to Split-2 and add this new one to allow negative axis!

File: docs/Changelog.md [hunk 14]
```diff
@@ -11424,6 +12376,207 @@ This version of the operator has been available since version 11 of the default
...
+#### Version
+
+This version of the operator has been available since version 11 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>axis</tt> : int (default is 0)</dt>
+<dd>Which axis to split on. A negative value means counting dimensions from the back. Accepted range is [-rank, rank-1].</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >rank [](start = 106, length = 4)

[-r, r-1] where r = rank(input).

File: docs/Changelog.md [hunk 15]
```diff
@@ -11424,6 +12376,207 @@ This version of the operator has been available since version 11 of the default
...
+
+#### Version
+
+This version of the operator has been available since version 11 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>axes</tt> : list of ints</dt>
+<dd>List of integers indicating the dimensions to squeeze. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 158, length = 5)

data

File: docs/Changelog.md [hunk 16]
```diff
@@ -11601,3 +12754,43 @@ This version of the operator has been available since version 11 of the default
...
+
+#### Version
+
+This version of the operator has been available since version 11 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>axes</tt> : list of ints (required)</dt>
+<dd>List of integers indicating the dimensions to be inserted. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input).</dd>
```
Comment (wschin) [inline docs/Changelog.md]: >input [](start = 162, length = 5)

data

File: docs/TestCoverage.md [hunk 17]
```diff
@@ -333,6 +357,30 @@ result = argmin_use_numpy(data, axis=axis, keepdims=keepdims)
...
+axis = -1
+keepdims = 1
+node = onnx.helper.make_node(
+    'ArgMin',
+    inputs=['data'],
+    outputs=['result'],
+    axis=axis,
+    keepdims=keepdims)
+# result: [[1], [0]]
+result = argmin_use_numpy(data, axis=axis, keepdims=keepdims)
```
Comment (wschin) [inline docs/TestCoverage.md]: >argmin_use_numpy [](start = 9, length = 16)

Your PR is already great. Just wondering if you can append something like
```
# The content of result is
# [1, 2, 3, 4]
```
to be more specific to the reader.

File: onnx/defs/math/defs.cc [hunk 18]
```diff
@@ -82,7 +83,29 @@ will throw errors.
...
+    schema.TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+      propagateElemTypeFromInputToOutput(ctx, 0, 0);
+      if (!hasNInputShapes(ctx, 1)) {
+        return;
+      }
+      propagateShapeFromInputToOutput(ctx, 0, 0);
+      const TensorShapeProto& input_shape =
+        ctx.getInputType(0)->tensor_type().shape();
+      int r = input_shape.dim_size();
+      if (r != 2) {
```
Comment (shinh) [inline onnx/defs/math/defs.cc]: I think this check does not follow the definition of `Softmax`? The document of `Softmax` says "Input does not need to explicitly be a 2D vector; rather, it will be coerced into one".
